design/file-descriptors.md

```diff
@@ -0,0 +1,17 @@
+# File Descriptors
+File Descriptors ([witx type `fd`](https://github.com/WebAssembly/WASI/blob/master/phases/snapshot/docs.md#-fd)) are an abstract way of representing access to a resource. More often than not, they are processes (WIP), sockets (WIP), files, or even the console.
+
+# Null File Descriptor
```

Comment:
0 is stdin, not null.

Reply:
Does a null `fd` exist?

Reply:
We don't have an answer for this yet :-}. POSIX often uses `-1` as an error code in functions that otherwise return a file descriptor. WASI APIs have a separate error return value, so they don't depend on special-case values like this. However, if we ever did return a file descriptor with this value, most C code would think it was an error.

Reply:
Note that Rust is in-progress defining `-1` as an invalid file descriptor, to the extent that assigning `-1` to a file descriptor type  would be undefined behavior: https://github.com/rust-lang/rust/pull/74699

This allows representing `Option<FileDesc>` with the layout of an `i32` and using the `-1` value to represent `None`.

Reply:
WASI has its own representation of `File` separate from the Unix definition, so that optimization as written doesn't apply to WASI. We could do the same optimization for WASI, but only if we decide to declare `-1` as invalid on WASI. On the surface it seems sensible; no typical host could support anywhere near a billion open files. And of course programs expecting POSIX semantics wouldn't be able to work with `-1`. My only concern here is in what happens as we generalize handles to support things other than files and high-level APIs other than POSIX and friends.

Reply:
In that case, I'll opt to leave out a null file descriptor section until more detail are finalized.